### PR TITLE
Fix #53: cover MoA boundary cases

### LINKED ISSUE #53: 2026犀牛鸟开源人才专属】【低难度】MoA(mixture of Attention)：边界测试补全与垂类训练验证
在现有 tests/test_moa.py 基础上，系统补全 MoA 模块的边界测试与训练验证，提升测试覆盖率
测试补全（至少覆盖以下场景）：
NeckMoAFusion 在跨尺度输入尺寸不匹配（如 hi 为 15×15、lo 为 7×7，非严格 2× 下采样）时的前向稳定性与形状保持
MoABlock 的 temperature 退火到极小值（如 temperature < 1e-4）时，softmax 路由概率的数值稳定性（是否出现 NaN 或均匀分布）
_LocalAttnHead 与 _GlobalAttnHead 在 num_heads 不能被 dim 整除时的降级处理（_safe_groups 的边界）
C2fMoA 的 aux_loss 在多 MoABlock 嵌套时是否存在重复计数（类似 MoE 的 MOE_LOSS_REGISTRY 双计数问题）
缺陷修复：在补充测试过程中，若发现任何边界缺陷（如 IndexError、NaN 传播、形状不匹配），需一并定位并修复
覆盖率报告：提供 pytest --cov 前后的覆盖率对比，至少覆盖 ultralytics/nn/modules/moa/ 目录
垂类训练验证：在 VisDrone 或 SKU-110K 上，使用 YOLO-Master-v0.10-MoA-N 训练 50~100 epoch，验证 MoA 模块在真实数据集上的收敛性，记录 mAP50-95 与 loss 曲线，与同配置的 MoE 基线对比
提交 Pull Request，包含测试代码、修复代码、覆盖率报告、训练日志
https://github.com/Tencent/YOLO-Master
https://github.com/Tencent/YOLO-Master/blob/main/ultralytics/nn/modules/moa/moa.py
https://github.com/Tencent/YOLO-Master/blob/main/tests/test_moa.py
https://github.com/Tencent/YOLO-Master/blob/main/ultralytics/cfg/models/master/v0_10/det/yolo-master-moa-n.yaml
https://github.com/Tencent/YOLO-Master/blob/main/scripts/compare_moa_ablation.py
难度：低

---

**本issue为2026犀牛鸟开源人才培养活动专属issue，仅供已报名参与犀牛鸟活动的同学认领**
【认领时间】7月1日～7月31日（7月1日前认领视为无效❗️）
【认领方式】在本issue评论区回复“已认领本任务”，即视为认领成功
【活动报名】需提前完成犀牛鸟报名问卷，问卷将用于活动登记和奖励发放：https://wj.qq.com/s2/26888567/gh2q

### PR DESCRIPTION
**Pull request by S1exy**
## 概要

Closes #53

本 PR 补齐 Issue #53 中 MoA 相关边界测试，并修复测试和训练过程中发现的问题。

主要包含：

- `NeckMoAFusion` 支持非严格 2x 尺度输入，例如 `15x15` 与 `7x7`
- `MoABlock` 在极小 annealed temperature，例如 `<1e-4`，下保持 router softmax 数值稳定
- `_LocalAttnHead` / `_GlobalAttnHead` 支持 `dim` 不能被 `num_heads` 整除的情况
- 修复 nested `C2fMoA` aux loss 重复计数问题
- 修复 Torch 2.0 SDPA 不支持 `scale=` 参数时的兼容性
- 移除 compare 脚本中重复的 MoA temperature annealing callback
- 补充 VisDrone 50 epoch baseline/MoA 对比训练报告和曲线图

## 测试

```bash
python -m pytest tests/test_issue53_training_scripts.py -q
python -m pytest tests/test_moa.py --cov=ultralytics/nn/modules/moa --cov-report=term-missing -q
python -m pytest tests/test_mixture_fixes.py tests/test_mixture_aux_loss.py -q
```

MoA 覆盖率：

```text
Before: tests/test_moa.py 7 passed, MoA module coverage 88%
After:  tests/test_moa.py 12 passed, MoA module coverage 94%
```

## 训练设置

数据集：VisDrone

硬件：RTX 3090 24GB

共同设置：

- `epochs=50`
- `batch=30`
- `imgsz=640`
- `workers=8`
- `cache=ram`
- `amp=False`
- `device=0`

baseline 和 MoA 都使用 no-AMP。原因是旧 AMP 训练排查中，NaN 首先出现在上游 `VisualEnhancedAdaptiveGateMoE`，而不是 MoA 模块本身；同一 MoA checkpoint/state 在 no-AMP 下可以稳定完成训练。因此本次公平对比对 baseline 和 MoA 同时关闭 AMP。

## 训练结果

Final epoch:

| Model | Epoch | Precision | Recall | mAP50 | mAP50-95 |
| --- | ---: | ---: | ---: | ---: | ---: |
| YOLO-Master v0.10 baseline | 50 | 0.32125 | 0.24908 | 0.22706 | 0.12492 |
| YOLO-Master v0.10 MoA | 50 | 0.32205 | 0.25108 | 0.22893 | 0.12574 |

Best checkpoint by mAP50-95:

| Model | Best Epoch | Precision | Recall | mAP50 | mAP50-95 |
| --- | ---: | ---: | ---: | ---: | ---: |
| YOLO-Master v0.10 baseline | 49 | 0.32251 | 0.24875 | 0.22778 | 0.12501 |
| YOLO-Master v0.10 MoA | 50 | 0.32205 | 0.25108 | 0.22893 | 0.12574 |

MoA 完整完成 50 epoch，并跨过旧 AMP run 中 epoch 14 附近的 NaN 失败点。训练日志中以下异常关键字匹配数为 0：

```text
TaskAlignedAssigner|OutOfMemory|Traceback|RuntimeError|nan|NaN|non-finite
```

## Loss 与指标曲线

Baseline:

YOLO-Master v0.10 baseline training curves

MoA:

YOLO-Master v0.10 MoA training curves

## 记录材料

- `scripts/issue53/coverage_report.md`
- `scripts/issue53/training_report_3090_noamp.md`
- `scripts/issue53/training_summary_3090_noamp.csv`
- `scripts/issue53/figures/v10_results.png`
- `scripts/issue53/figures/v10_moa_results.png`

本 PR 未提交数据集、模型权重、完整训练日志或 cache 文件。

### CODE REVIEW

**Comment by isLinXu** [comment]
LGTM
